### PR TITLE
Expand other branch pipes

### LINKED ISSUE #415: RPC docking APi gateway
Excuse me. We have a requirement that requires RPC to also connect to the APi gateway. Is there a corresponding document?

### PR DESCRIPTION
**Pull request by lfzm**
## Closes #415 
- #415

### CODE REVIEW

**Comment by TomPallister** [comment]
@aqa510415008 this is cool! We do need unit tests for it in case anyone decides to change your logic and breaks it. Please aim for 100% code coverage.

**Comment by TomPallister** [comment]
#415

**Comment by aqa510415008** [comment]
@TomPallister  Unit test has been added, trouble review

**Comment by TomPallister** [comment]
@aqa510415008 will review ASAP! Thank you 😄

**Comment by aqa510415008** [comment]
@TomPallister I hope to add two attributes to `IOcelotBuilder `  for easy extension.
```
        IServiceCollection Services { get; }
        IConfiguration Configuration { get; }
```

File: src/Ocelot/DependencyInjection/OcelotBuilder.cs [hunk 1]
```diff
@@ -55,6 +55,10 @@ public class OcelotBuilder : IOcelotBuilder
         private readonly IServiceCollection _services;
         private readonly IConfiguration _configurationRoot;
 
+        public IServiceCollection Services => _services;
```
Comment (TomPallister) [inline src/Ocelot/DependencyInjection/OcelotBuilder.cs]: Why do we need to do this? People can already edit these outside of Ocelot builder if they want to.

File: src/Ocelot/DependencyInjection/IOcelotBuilder.cs [hunk 2]
```diff
@@ -4,11 +4,15 @@
...
 using IdentityServer4.AccessTokenValidation;
 using Ocelot.Middleware.Multiplexer;
+using Microsoft.Extensions.DependencyInjection;
+using Microsoft.Extensions.Configuration;
 
 namespace Ocelot.DependencyInjection
 {
     public interface IOcelotBuilder
     {
+        IServiceCollection Services { get; }
```
Comment (TomPallister) [inline src/Ocelot/DependencyInjection/IOcelotBuilder.cs]: Why do we need to do this? People can already edit these outside of Ocelot builder if they want to.

**Review by TomPallister — CHANGES_REQUESTED** [review]
The code looks good. Please can you look at my comment. I do not understand why we are exposing services and configuration.

Also please could you provide a small piece of documentation on how to use this feature? It doesnt have to be in english I can translate using google.

**Comment by aqa510415008** [comment]
@TomPallister  

MapWhenOcelotPipeline - Customize other pipelines and connect other RPCs to Ocelot, such as GRPC, Orleans, and so on.

## Simple WebSocket pipe extension
```
    OcelotPipelineConfiguration config = new OcelotPipelineConfiguration();
    config.MapWhenOcelotPipeline.Add(builder =>
    {
        builder.UseDownstreamRouteFinderMiddleware();
        builder.UseDownstreamRequestInitialiser();
        builder.UseLoadBalancingMiddleware();
        builder.UseDownstreamUrlCreatorMiddleware();
        builder.UseWebSocketsProxyMiddleware();
        //Return function to determine whether to enter this pipeline
        return (DownstreamContext context) => context.HttpContext.WebSockets.IsWebSocketRequest;
    });
```

File: src/Ocelot/DependencyInjection/IOcelotBuilder.cs (hunk 2 above)
Comment (TomPallister) [reply src/Ocelot/DependencyInjection/IOcelotBuilder.cs]: Ok I understand

**Comment by TomPallister** [comment]
@aqa510415008 I will merge this soon! Thanks for your contribution it is appreciated!

**Comment by aqa510415008** [comment]
@TomPallister 
I have implemented the convergence of `Ocelot ` and  `Orleans ` here, and `Ocelot ` is the Api gateway of 
 `Orleans` .
 [Ocelot.OrleansHttpGateway](https://github.com/aqa510415008/Ocelot.OrleansHttpGateway)

File: src/Ocelot/DependencyInjection/OcelotBuilder.cs (hunk 1 above)
Comment (geffzhang) [reply src/Ocelot/DependencyInjection/OcelotBuilder.cs]: the extended components can be configured under Ocelot ,@aqa510415008  have implemented the convergence of Ocelot and Orleans here, and Ocelot is the Api gateway of
Orleans .
Ocelot.OrleansHttpGateway

**Review by geffzhang — APPROVED** [review]
This implementation is very good

**Review by TomPallister — APPROVED** [review]

**Pull request merged**